### PR TITLE
add blocked version to address performance issue of when N is large

### PR DESCRIPTION
**Pull request by xiaohuguo2023**
blocked rmsnorm implementation
-  benefit when N >30K, the best BLOCK_SIZE = 65536
-  add unit test for N > 30K
-  add the logic in the wrapper to set the best BLOCK_SIZE
- for non block version, this pr won't have any impart of performance.
- non-blocked v.s. blocked version perf comparison

|       N   | Triton (non blocked) | Triton (blocked) | Improvement (%) |
|-----------|--------------|--------------|-----------------|
|   31744.0 |     11.736   |    11.615    |      -1.03      |
|   41984.0 |      7.326   |    11.911    |      62.59      |
|   52224.0 |      8.381   |    13.114    |      56.42      |
|   62464.0 |      9.721   |    15.099    |      55.30      |
|   72704.0 |      1.090   |    14.699    |    1249.45      |
|   82944.0 |      1.272   |    15.360    |    1107.08      |
|   93184.0 |      1.431   |    16.568    |    1058.33      |
|  103424.0 |      1.614   |    16.212    |     904.21      |
|  113664.0 |      1.766   |    18.606    |     953.76      |
|  123904.0 |      1.896   |    18.398    |     870.12      |
|  134144.0 |      1.284   |    17.659    |    1275.78      |
|  144384.0 |      1.531   |    17.966    |    1073.66      |
|  154624.0 |      1.636   |    19.041    |    1063.99      |
|  164864.0 |      1.578   |    18.224    |    1054.63      |
|  175104.0 |      1.852   |    19.307    |     942.62      |
|  185344.0 |      1.951   |    19.420    |     895.32      |
|  195584.0 |      1.894   |    19.194    |     914.01      |
|  205824.0 |      1.953   |    19.811    |     914.43      |
|  216064.0 |      2.071   |    19.505    |     841.59      |
|  226304.0 |      2.163   |    19.914    |     820.58      |
|  236544.0 |      2.234   |    18.855    |     743.71      |
|  246784.0 |      2.557   |    21.409    |     737.61      |

### CODE REVIEW

File: python/perf-kernels/rmsnorm.py [hunk 1]
```diff
@@ -37,41 +37,105 @@ def get_autotune_config():
...
+            output_row_start_ptr = output_ptr + row_idx * output_row_stride
+            output_ptrs = output_row_start_ptr + col_offsets
+            output_ptrs = tl.multiple_of(output_ptrs, (16, ))
+            tl.store(output_ptrs, rms_norm, mask=mask)
+
+
 def triton_rmsnorm(x, y, g, n_rows, n_cols, blk_size, epsilon=1e-6):
     BLOCK_SIZE = blk_size
+    # Use blocked approach if BLOCK_SIZE > 65536
+    USE_BLOCKED = BLOCK_SIZE > 31743
```
Comment (rahulbatra85) [inline python/perf-kernels/rmsnorm.py]: One thing that I noticed in the layernorm tutorial was also using the dtype to determine whether to use blocked or non-blocked. I think what matters is not the actual number of elements, but the total size of elements.

Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py]: you are right, line 131 should be
```
if n_cols > 65535
```

Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py]: done

Comment (rahulbatra85) [reply python/perf-kernels/rmsnorm.py]: Actually, I was suggesting more like this 
https://github.com/ROCm/triton/blob/main_perf/python/perf-kernels/layernorm.py#L118

**Comment by vgokhale** [comment]
Can you please add performance comparison with and without this change?

**Comment by xiaohuguo2023** [comment]
> Can you please add performance comparison with and without this change?



> Can you please add performance comparison with and without this change?

see above

File: python/perf-kernels/rmsnorm.py (hunk 1 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py]: revised as suggested.

File: python/perf-kernels/rmsnorm.py [hunk 2]
```diff
@@ -37,41 +37,103 @@ def get_autotune_config():
...
+
+        # Apply normalization
+        apply_normalization(input_ptr, output_ptr, g_ptr, input_row_stride, output_row_stride, n_cols, norm_factor,
+                            BLOCK_SIZE, row_start)
+
+    else:
+        mask = col_offsets < n_cols
+        tl.assume(input_row_stride >= 0)
+        tl.assume(output_row_stride >= 0)
+        for row_idx in tl.range(row_start, n_rows, NUM_PRGMS):
```
Comment (vgokhale) [inline python/perf-kernels/rmsnorm.py]: Is this loop needed? NUM_PRGMS = n_rows in the caller. So it will never execute more than once?

Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py]: yes, if N is small, and M > 304, we will need this persistent loop

Comment (vgokhale) [reply python/perf-kernels/rmsnorm.py]: In persistent kernel the grid is sized according to the number of CUs.

On line 134 and 135, the grid is set as number of rows. It is agnostic to CUs.

How is this kernel persistent?

Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py]: you are correct, this is a fake persistent loop, I will remove it, and to get it work for persistent, I need a new PR for this

**Review by scxiao — APPROVED** [review]

File: python/perf-kernels/rmsnorm.py [hunk 3]
```diff
@@ -153,7 +217,8 @@ def benchmark(M, N, provider):
         x = torch.randn(M, N, device='cuda', dtype=dtype)
         y = torch.zeros_like(x, device='cuda')
         n_rows, n_cols = x.shape
-        blk_size = triton.next_power_of_2(n_cols)
+        MAX_FUSED_SIZE = 65536 // x.element_size()
```
Comment (vgokhale) [inline python/perf-kernels/rmsnorm.py:220]: Can you add some comments on this magic number?

**Comment by vgokhale** [comment]
LGTM after outstanding comments are addressed.

File: python/perf-kernels/rmsnorm.py [hunk 4]
```diff
@@ -37,28 +37,87 @@ def get_autotune_config():
...
+    row_output_ptr = output_ptr + row_idx * output_row_stride
+
+    for start in range(0, n_cols, BLOCK_SIZE):
+        cols = start + col_offsets
+        mask = cols < n_cols
+        input_ptrs = row_input_ptr + cols
+        input_ptrs = tl.multiple_of(input_ptrs, (16, ))
+        g_ptrs = g_ptr + cols
+        output_ptrs = row_output_ptr + cols
+        x = tl.load(input_ptrs, mask=mask, other=0.0, cache_modifier=".cg")
```
Comment (vgokhale) [inline python/perf-kernels/rmsnorm.py:81]: Have you tried peeling the last iter? is it worth trying? Can you add a TODO to try that as part of your next PR?

File: python/perf-kernels/rmsnorm.py (hunk 3 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py:220]: sorry, just saw your messages, anything great than MAX_FUSED_SIZE, we start to have spills

File: python/perf-kernels/rmsnorm.py (hunk 4 above)
Comment (xiaohuguo2023) [reply python/perf-kernels/rmsnorm.py:81]: the loop peeling only applied for blocked version, not this one. sure, I will try that

**Pull request merged**